File: white-hole/src/Operations/InterfaceOperations.hs
{-# LANGUAGE OverloadedStrings #-}
module Operations.InterfaceOperations where

import Database.PostgreSQL.Simple
import Entities.User
import qualified Entities.Movie as M
import Entities.Serie
import Data.Char (toLower, toUpper, isNumber, isAlphaNum, isDigit)
import qualified Entities.Rating as R
import Operations.UserOperations
import Operations.MovieOperations
import Operations.MovieInterfaceOperations
import Operations.SerieInterfaceOperations
import Operations.UtilOperations
import System.Exit
import System.IO
import Operations.SerieOperations
import qualified System.Process as System


signUp :: Connection -> IO()
signUp conn = do
    putStrLn ""
    putStrLn "Digite o seu nome:"
    userName <- getLine
    if null (strip userName) then do 
        putStrLn "Digite um nome válido" 
        clearScreenWithConfirmation
        return () 
    else do 
        putStrLn ""

    putStrLn "Digite o seu sobrenome:"
    userSurname <- getLine
    if null (strip userSurname) then do 
        putStrLn "Digite um sobrenome válido" 
        clearScreenWithConfirmation
        return ()
    else do 
        putStrLn ""

    putStrLn "Digite o seu e-mail:"
    userEmail <- getLine
    if null (strip userEmail) then do 
        putStrLn "Digite um e-mail válido" 
        clearScreenWithConfirmation
        return ()
    else do 
        putStrLn ""

    putStrLn "Digite a sua senha (tem que ter no mínimo 8 caracteres):"
    userPassword <- getLine
    if length (strip userPassword) < 9 then do 
        putStrLn "A senha não entende aos requisitos básicos!"
        clearScreenWithConfirmation
        return ()
    else do
        putStrLn ""
    ok <- createUser conn $ User {email=userEmail, password=userPassword, nome=userName, sobrenome=userSurname}
    putStrLn (if ok then "Usuario cadastrado com sucesso!" else "Ja existe usuario cadastrado com esse e-mail!")
    clearScreenWithConfirmation


signIn :: Connection -> IO()
signIn conn = do
    putStrLn ""
    putStrLn "Digite o seu e-mail:"
    userEmail <- getLine
    putStrLn "Digite a sua senha:"
    userPassword <- getLine
    {- Chamada da função run que roda as opções de avaliação e de 
    pesquisa, assim como os rankings! -}
    authenticated <- authenticate conn userEmail userPassword
    if authenticated then do
        putStrLn ""
        putStrLn "Credenciais válidas!"
        user <- getUserByEmail conn userEmail
        clearScreenWithConfirmation 
        run conn $ head user
    else do
        putStrLn ""
        putStrLn "E-mail ou senha inválidos!"
        clearScreenWithConfirmation
        


firstMenu :: Connection -> IO()
firstMenu conn = do
    putStrLn ""
    putStrLn "Bem vindo ao White Hole de filmes e séries!"
    putStrLn "Descubra novos filmes para assistir e avalie os que você já assistiu!"
    putStrLn ""
    putStrLn "1 - Logar no White Hole"
    putStrLn "2 - Cadastrar-se no White Hole"
    putStrLn "3 - Sair"
    putStrLn ""
    putStrLn "Digite a opção que deseja acessar:"
    option <- getLine
    if option == "1" then do
        clearScreenOnly >> signIn conn >> firstMenu conn
    else if option == "2" then do
        clearScreenOnly >> signUp conn >> firstMenu conn
    else if option == "3" then do
        putStrLn ""
        putStrLn "Obrigado, tenha um ótimo dia!"
        putStrLn ""
    else do
        putStrLn ""
        putStrLn "Digite uma opção válida na próxima!"
        putStrLn ""
        clearScreenWithConfirmation


run :: Connection -> User -> IO()
run conn user = do
    putStrLn ("Olá " ++ nome user ++ "!")
    putStrLn ""
    putStrLn "1 - Procurar por filme ou série"
    putStrLn "2 - Minha lista de marcados para assistir depois"
    putStrLn "3 - 10 melhores filmes"
    putStrLn "4 - 10 melhores séries"
    putStrLn "5 - 10 melhores filmes por categoria"
    putStrLn "6 - 10 melhores séries por categoria"
    putStrLn "7 - Recomendações para mim"
    putStrLn "8 - Logout"
    putStrLn ""
    putStrLn "Digite a opção desejada: "
    hFlush stdout
    option <- getLine
    clearScreenOnly
    case option of
        "1" -> clearScreenOnly >> search conn user >> run conn user
        "2" -> clearScreenOnly >> myList conn user >> run conn user
        "3" -> clearScreenOnly >> tenBestMovies conn user >> run conn user
        "4" -> clearScreenOnly >> tenBestSeries conn user >> run conn user
        "5" -> clearScreenOnly >> tenBestMoviesByCategory conn user >> run conn user
        "6" -> clearScreenOnly >> tenBestSeriesByCategory conn user >> run conn user
        "7" -> clearScreenOnly >> recomendations conn user >> run conn user
        "8" -> putStrLn "" >> putStrLn "Até mais, volte sempre!"
        x -> putStrLn "Digite uma opção válida" >> clearScreenWithConfirmation >> run conn user


search :: Connection -> User -> IO()
search conn user = do
    putStrLn "---------------------- PESQUISAR ------------------------"
    putStrLn ""
    putStrLn "Digite o nome do filme ou série que deseja pesquisar (Se "
    putStrLn "quiser voltar, aperte enter):"
    putStrLn ""
    title <- getLine
    if null (strip title) then do 
        clearScreenOnly
    else do
        putStrLn ""
        movies <- searchMovie conn (map toLower title)
        series <- searchSerie conn (map toLower title)
        if null movies && null series then do
            putStrLn "Não foi encontrado nenhum filme ou série com esse título!" 
        else do
            putStrLn "Filmes:"
            putStrLn ""
            printMoviesList movies 1
            putStrLn ""
            putStrLn "Séries:"
            putStrLn ""
            printSeriesList series (fromIntegral (length movies) + 1)

        putStrLn ""
        putStrLn "Digite o número correspondente ao filme ou a série que quer ver ou as opções abaixo:"
        putStrLn "( p - Pesquisar novamente; v - Voltar ao menu principal; c - Cadastrar um novo filme ou série): "
        putStrLn ""

        option <- getLine
        if option == "v" then
            clearScreenOnly
        else if option == "p" then
            clearScreenOnly >> search conn user
        else if option == "c" then
            clearScreenOnly >> registerMovieSerie conn user
        else if not (isANumber option True) || (((read option :: Int) > (length movies + length series)) || (read option :: Int) < 1) || null option then do
            putStrLn "Digite uma opção válida na próxima vez." >> clearScreenWithConfirmation
        else do
            if (read option :: Int) <= length movies then do
                clearScreenOnly >> showMovie conn user (movies !! ((read option :: Int) - 1)) >> search conn user
            else do
                clearScreenOnly >> showSerie conn user (series !! ((read option :: Int) - length movies - 1)) >> search conn user


registerMovieSerie :: Connection -> User -> IO ()
registerMovieSerie conn user = do
    putStrLn ""
    putStrLn "1 - Adicionar filme"
    putStrLn "2 - Adicionar série"
    putStrLn "v - Voltar"
    putStrLn ""
    putStrLn "Digite a opção:"
    option <- getLine
    case option of
        "1" -> clearScreenOnly >> createMovie conn >> search conn user
        "2" -> clearScreenOnly >> createSerie conn >> search conn user
        "v" -> clearScreenOnly >> search conn user
        x -> putStrLn "Digite uma opção válida!" >> clearScreenWithConfirmation >> registerMovieSerie conn user


myList :: Connection -> User -> IO ()
myList conn user = do
    putStrLn "------ Minha Lista Para Assistir Depois ------"
    putStrLn ""
    movies <- getWatchLaterList conn user
    series <- getWatchLaterListSeries conn user
    putStrLn "Filmes:"
    putStrLn ""
    if null movies then do
        print "Lista vazia"
    else do
        printMoviesList movies 1

    putStrLn ""
    putStrLn "Séries:"
    putStrLn ""
    
    if null series then do
        print "Lista vazia"
    else do
        printSeriesList series (fromIntegral (length movies) + 1)

    putStrLn ""
    putStrLn "Opções:"
    putStrLn ""
    putStrLn "1 - Acessar filme ou série"
    putStrLn "2 - Voltar"
    putStrLn ""
    putStrLn "Digite sua ação"
    option <- getLine
    if option == "1" then do
        putStrLn ""
        putStrLn "Qual filme ou série você quer acessar? (Digite o número do filme/série da lista; Para voltar, aperte Enter)"
        choice <- getLine
        if null choice then do 
            clearScreenOnly 
        else do 
            if (read choice :: Int) > length movies + length series || (read choice :: Int) <= 0 then do
                putStrLn ""
                putStrLn "digite uma opção válida" >> clearScreenWithConfirmation
                myList conn user
            else do
                if (read choice :: Int) <= length movies then do
                    clearScreenOnly >> showMovie conn user (movies !! ((read choice :: Int) - 1)) >> myList conn user
                else do
                    clearScreenOnly >> showSerie conn user (series !! ((read choice :: Int) - length movies - 1)) >> myList conn user

    else do
        if option == "2" then do
            clearScreenOnly
        else do
            putStrLn ""
            putStrLn "Digite uma opção válida" >> clearScreenWithConfirmation
            myList conn user


recomendations :: Connection -> User -> IO ()
recomendations conn user = do
    putStrLn ""
    putStrLn "--------------------------------------------------------"
    putStrLn "Recomendações"
    putStrLn "--------------------------------------------------------"
    putStrLn ""
    putStrLn "Aqui estão algumas recomendações de filmes e séries que "
    putStrLn "achamos que você vai gostar!"
    putStrLn ""
    putStrLn "Analizando o seu perfil... (aguarde)"
    putStrLn ""
    movies <- getRecomendationsOfMovies conn user
    series <- getRecomendationsOfSeries conn user
    putStrLn "Filmes:"
    putStrLn ""
    if null movies then do
        tenMovies <- getTenBestMovies conn 
        printMoviesList tenMovies 1 
    else do
        printMoviesList movies 1
    putStrLn ""
    putStrLn "Séries:"
    putStrLn ""
    if null series then do 
        tenSeries <- getTenBestSeries conn
        printSeriesList tenSeries (fromIntegral (length movies + 1))
    else do 
        printSeriesList series (fromIntegral (length movies + 1))
    putStrLn ""
    putStrLn "Qual filme ou série você quer acessar? (Digite o número do filme/série da lista; Se quiser voltar, aperte Enter)"
    choice <- getLine
    if null choice then do
        clearScreenOnly
    else do
        if (read choice :: Int) > length movies + length series || (read choice :: Int) <= 0 then do
            putStrLn ""
            putStrLn "digite uma opção válida" >> clearScreenWithConfirmation
            recomendations conn user
        else do
            if (read choice :: Int) <= length movies then do
                clearScreenOnly >> showMovie conn user (movies !! ((read choice :: Int) - 1)) >> recomendations conn user
            else do
                clearScreenOnly >> showSerie conn user (series !! ((read choice :: Int) - length movies - 1)) >> recomendations conn user

File: white-hole-pl/src/Operations/interface_operations.pl
:- module(interfaceop, [first_menu/1]).
:- use_module('./util.pl', [get_input/2, clear_screen/0, clear_screen_with_confirmation/0]).
:- use_module('./user_operations.pl').
:- use_module('./movie_operations.pl').
:- use_module('./serie_operations.pl').
:- use_module('./serie_interface_operations.pl').
:- use_module('./movie_interface_operations.pl').


first_menu(Connection) :-
    writeln(''),
    writeln('Bem vindo ao White Hole de filmes e séries!'),
    writeln('Descubra novos filmes para assistir e avalie os que você já assistiu!'),
    writeln(''),
    writeln('1 - Logar no White Hole'),
    writeln('2 - Cadastrar-se no White Hole'),
    writeln('3 - Sair'),
    writeln(''),
    get_input('Digite a opção que deseja acessar:', Option),
    (Option = "1" -> 
        clear_screen, signin(Connection), first_menu(Connection);
        (Option = "2" ->
            clear_screen, signup(Connection), first_menu(Connection);
            (Option = "3" ->
                clear_screen, writeln('Obrigado, tenha um ótimo dia!'), clear_screen_with_confirmation, halt;
                (Option = _ ->
                    clear_screen, writeln('Digite uma opção válida na próxima!'), writeln(''), clear_screen_with_confirmation, first_menu(Connection);
                    !
                )           
            )    
        )
    ).


signin(Connection) :-
    writeln(''),
    get_input('Digite o seu e-mail:', UserEmail),
    writeln(''),
    get_input('Digite a sua senha:', UserPassword),
    clear_screen,
    authenticate(Connection, UserEmail, UserPassword, Autenticated),
    (Autenticated =:= 1 ->
        getUserByEmail(Connection, UserEmail, [User]),
        run(Connection, User);
        writeln(''),
        writeln('E-mail ou senha inválidos!'),
        clear_screen_with_confirmation  
    ).

signup(Connection) :-
    writeln(''),
    get_input('Digite o seu nome:', UserName),
    (UserName = "" ->
         writeln('Digite um nome valido'),
         signup(Connection);
         !
    ),
    writeln(''),
    get_input('Digite o seu sobrenome:', UserSurName),
    (UserSurName = "" ->
         writeln('Digite um sobrenome valido'),
         signup(Connection);
         !
    ), 
    writeln(''),
    get_input('Digite um E-mail:', UserEmail),
    (UserEmail = "" ->
         writeln('Digite um E-mail valido'),
         signup(Connection);
         !
    ),
    writeln(''),
    get_input('Digite a sua senha (tem que ter no mínimo 8 caracteres):', UserPassword),
    (UserPassword = "" ->
         writeln('Digite uma senha valida'),
         signup(Connection);
         !
    ), 
    createUser(Connection, UserEmail, UserPassword, UserName, UserSurName, Confirmacao),
    (Confirmacao =:= 1 ->
         writeln('Usuario cadastrado com sucesso!'),
         clear_screen_with_confirmation,
         first_menu(Connection);
         writeln('Usuario não cadastrado!'),
         clear_screen_with_confirmation,
         first_menu(Connection)
    ).


run(Connection, User):-
    writeln('Olá!'),
    writeln(''),
    writeln('1 - Procurar por filme ou série'),
    writeln('2 - Minha lista de marcados para assistir depois'),
    writeln('3 - 10 melhores filmes'),
    writeln('4 - 10 melhores séries'),
    writeln('5 - 10 melhores filmes por categoria'),
    writeln('6 - 10 melhores séries por categoria'),
    writeln('7 - Recomendações para mim'),
    writeln('8 - Logout'),
    writeln(''),
    get_input('Digite a opção desejada: ', Option),
    (
        Option = "1" -> clear_screen, search(Connection, User), run(Connection, User);
        Option = "2" -> clear_screen, myList(Connection, User), run(Connection, User);
        Option = "3" -> clear_screen, tenBestMovies(Connection, User), run(Connection, User);
        Option = "4" -> clear_screen, serie_interface_operations:tenBestSeries(Connection, User), run(Connection, User);
        Option = "5" -> clear_screen, tenBestMoviesByCategory(Connection, User), run(Connection, User);
        Option = "6" -> clear_screen, serie_interface_operations:tenBestSeriesByCategory(Connection, User), run(Connection, User);
        Option = "7" -> clear_screen, recomendations(Connection, User), run(Connection, User);
        Option = "8" -> writeln(""), writeln("Até mais, volte sempre!");
            writeln("Digite uma opção válida"), clear_screen_with_confirmation, run(Connection, User)
    ).


search(Connection, User):-
    writeln('---------------------- PESQUISAR ------------------------'),
    writeln(""),
    writeln("Digite o nome do filme ou série que deseja pesquisar (Se "),
    writeln("quiser voltar, aperte enter):"),
    get_input("", Title),
    util:strip(Title, TitleTrimmed),
    (TitleTrimmed = "" -> clear_screen ;
        writeln(""),
        string_lower(TitleTrimmed, TitleForSearch),
        searchMovie(Connection, TitleForSearch, Movies),
        searchSerie(Connection, TitleForSearch, Series),
        ( Movies = [], Series = [] -> 
            writeln("Não foi encontrado nenhum filme ou série com esse título!" );
            writeln("Filmes:"),
            writeln(""),
            printMoviesList(Movies, 1),
            writeln(""),
            writeln("Séries:"),
            writeln(""),
            length(Movies, MLength),
            SeriesNum is MLength + 1,
            printSeriesList(Series, SeriesNum)
            ),
        writeln(""),
        writeln("Digite o número correspondente ao filme ou a série que quer ver ou as opções abaixo:"),
        writeln("( p - Pesquisar novamente; v - Voltar ao menu principal; c - Cadastrar um novo filme ou série): "),
        get_input("", Option),
        ( Option = "v" -> clear_screen;
            Option = "p" -> clear_screen, search(Connection, User);
            Option = "c" -> clear_screen, registerMovieSerie(Connection, User);
            length(Movies, MLength), length(Series, SLength), Total is MLength + SLength,
            util:isANumber(NOption, Option), NOption =< Total, NOption > 0, Option \= "" ->
                ( NOption =< MLength -> 
                    nth1(NOption, Movies, Movie),
                    clear_screen, showMovie(Connection, User, Movie),search(Connection, User)
                    ;
                    SIndex is NOption - MLength,
                    nth1(SIndex, Series, Serie),
                    clear_screen, showSerie(Connection, User, Serie),search(Connection, User)
                )
                ;
                writeln("Digite uma opção válida na próxima vez."), clear_screen_with_confirmation    
            )
        ).


registerMovieSerie(Connection, User):-
    writeln(""),
    writeln("1 - Adicionar filme"),
    writeln("2 - Adicionar série"),
    writeln("v - Voltar"),
    writeln(""),
    get_input("Digite a opção:", Option),
    (
        Option = "1" -> clear_screen, createMovie(Connection), search(Connection, User);
        Option = "2" -> clear_screen, createSerie(Connection), search(Connection, User);
        Option = "v" -> clear_screen, search(Connection, User);
            writeln("Digite uma opção válida!"), clear_screen_with_confirmation, registerMovieSerie(Connection, User)
    ).


myList(Connection, User):-
    writeln("------ Minha Lista Para Assistir Depois ------"),
    writeln(""),
    movie_operations:getWatchLaterList(Connection, User, Movies),
    serie_operations:getWatchLaterListSeries(Connection, User, Series),
    writeln("Filmes:"),
    writeln(""),
    (Movies = [] -> writeln("Lista vazia"); printMoviesList(Movies, 1)),
    
    writeln(""),
    writeln("Séries:"),
    writeln(""),
    SL is MLength + 1,
    (Series = [] -> writeln("Lista vazia"); length(Movies, MLength), printSeriesList(Movies, SL)),

    writeln(""),
    writeln("Opções:"),
    writeln(""),
    writeln("1 - Acessar filme ou série"),
    writeln("2 - Voltar"),
    writeln(""),
    get_input("Digite sua ação", Option),
    ( Option = "1" ->
        writeln(""),
        get_input("Qual filme ou série você quer acessar? (Digite o número do filme/série da lista; Para voltar, aperte Enter)", Choice),
        ( Choice = "" -> 
            clear_screen;
            length(Movies, MLength), length(Series, SLength),
            ( util:isANumber(NChoice, Choice), NChoice < MLength + SLength ->
                ( NChoice > 1 -> 
                    ( NChoice =< MLength ->
                        nth1(NChoice, Movies, Movie),
                        clear_screen, 
                        showMovie(Connection, User, Movie), myList(Connection, User);
                        nth1(NChoice - MLength, Series, Serie),
                        clear_screen,
                        showSerie(Connection, User, Serie), myList(Connection, User)
                    )
                    ;
                    writeln(""),
                    writeln("digite uma opção válida"), clear_screen_with_confirmation,
                    myList(Connection, User)
                );
                writeln(""),
                writeln("digite uma opção válida"), clear_screen_with_confirmation,
                myList(Connection, User)
            )
        );
    Option = "2" -> clear_screen;
        writeln(""),
        writeln("Digite uma opção válida"), clear_screen_with_confirmation,
        myList(Connection, User)
    ).

recomendations(Connection, User):-
    writeln(""),
    writeln("--------------------------------------------------------"),
    writeln("Recomendações"),
    writeln("--------------------------------------------------------"),
    writeln(""),
    writeln("Aqui estão algumas recomendações de filmes e séries que "),
    writeln("achamos que você vai gostar!"),
    writeln(""),
    writeln("Analizando o seu perfil... (aguarde)"),
    writeln(""),
    movie_operations:getRecomendationsOfMovies(Connection, User, Movies),
    serie_operations:getRecomendationsOfSeries(Connection, User, Series),
    writeln("Filmes:"),
    writeln(""),
    getTenBestMovies(Connection, TenBestMovies),
    ( Movies = [] -> length(TenBestMovies, MLength), printMoviesList(TenBestMovies, 1); length(Movies, MLength), printMoviesList(Movies, 1) ),
    writeln(""),
    writeln("Séries:"),
    writeln(""),
    getTenBestSeries(Connection, TenBestSeries),
    ( Series = [] -> length(TenBestSeries, SLength), printMoviesList(TenBestSeries, 1); length(Series, SLength), MSL is MLength + 1, printSeriesList(Series, MSL) ),
    writeln(""),
    get_input("Qual filme ou série você quer acessar? (Digite o número do filme/série da lista; Se quiser voltar, aperte Enter)", Choice),
    ( Choice = "" -> 
            clear_screen;
            length(Movies, MLength), length(Series, SLength),
            ( util:isANumber(NChoice, Choice), NChoice < MLength + SLength ->
                ( NChoice > 1 -> 
                    ( NChoice =< MLength ->
                        nth1(NChoice, Movies, Movie),
                        clear_screen, 
                        showMovie(Connection, User, Movie), recomendations(Connection, User);
                        nth1(NChoice - MLength, Series, Serie),
                        clear_screen,
                        showSerie(Connection, User, Serie), recomendations(Connection, User)
                    )
                    ;
                    writeln(""),
                    writeln("digite uma opção válida"), clear_screen_with_confirmation,
                    recomendations(Connection, User)
                );
                writeln(""),
                writeln("digite uma opção válida"), clear_screen_with_confirmation,
                recomendations(Connection, User)
            )
    ).